src/java.base/share/classes/com/sun/crypto/provider/ChaCha20Cipher.java

```diff
@@ -753,6 +769,120 @@ protected int engineDoFinal(byte[] in, int inOfs, int inLen, byte[] out,
         return bytesUpdated;
     }
 
+    /**
+     * Complete the currently running operation using any final
+     * data provided by the caller.
+     *
+     * @param input the plaintext or ciphertext input bytebuffer.
+     * @param output ByteBuffer that will hold the resulting data.  This
+     *      must be large enough to hold the resulting data.
+     *
+     * @return the resulting plaintext or ciphertext bytes.
+     *
+     * @throws AEADBadTagException if, during decryption, the provided tag
+     *      does not match the calculated tag.
+     */
+    @Override
+    protected int engineDoFinal(ByteBuffer input, ByteBuffer output)
+        throws ShortBufferException {
+        return bufferCrypt(input, output, false);
+    }
+
+    /*
+     * Optimized version of bufferCrypt from CipherSpi.java.  Direct
```

Comment:
well I was under the impression that the bulk of the optimization comes from using modified `engineGetOutputSize` that does not allocate when isUpdate== true and decrypting. That's not documented anywhere.

Reply:
It used the `engineGetOutputSize` from ChaCha20Cipher before the change.  The biggest gain no doing unnecessary copying for direct bytebuffers.  For a decrypt op, it would copy the bytebuffer into and array, then copy the array into the internal buffer.   I try to minimize that by doing only one copy, bytebuffer -> internal buffer.